subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/metadata/AbstractRepositoryMetadataSource.java

```diff
@@ -79,12 +85,26 @@ private S parseMetaDataFromArtifact(String repositoryName, ComponentResolvers co
                         throw new IllegalStateException("Unexpected Gradle metadata redirection answer");
                     }
                 }
-                return parseResult.getResult();
+                S metadata = parseResult.getResult();
+                File metadataArtifactFile = metadataArtifact.getFile();
+                ExternalResourceMetaData metaData = metadataArtifact.getMetaData();
+                MutableModuleSources sources = metadata.getSources();
+                sources.add(new DefaultMetadataFileSource(artifact.getId(), metadataArtifactFile, findSha1(metaData, metadataArtifactFile)));
+                context.appendSources(sources);
+                return metadata;
             }
         }
         return null;
     }
 
+    static BigInteger findSha1(ExternalResourceMetaData metaData, File artifact) {
+        HashValue sha1 = metaData.getSha1();
```

Comment:
Out of curiosity, and maybe I just missed it: What happens if the published sha is wrong? What kind of error will the use see at which moment?

Reply:
Here I think it just means the file would be stored in the "wrong" directory.